From 112940df9a08e25eaab2dc5d263cfdcdd90b06d5 Mon Sep 17 00:00:00 2001
From: Deepak Rathore <deepak.rathore@einfochips.com>
Date: Mon, 9 May 2022 17:42:51 +0530
Subject: [PATCH] IMX8M: Added changes to make Auto Focus functionality working

- Set Actuator absolute milisecond time to 5ms
- Initialise actuator controls ACT_CTRL_0 and ACT_CTRL_1 values

Signed-off-by: Deepak Rathore <deepak.rathore@einfochips.com>

diff --git a/drivers/media/i2c/ap1302.c b/drivers/media/i2c/ap1302.c
index f86731300..469d12db9 100644
--- a/drivers/media/i2c/ap1302.c
+++ b/drivers/media/i2c/ap1302.c
@@ -388,6 +388,9 @@
 #define AWB_MODE_0				0x00000000
 #define AWB_MODE_1				0x00000001
 #define AF_MODE_1				0x0001
+#define ACT_ABS_MTIME				AP1302_REG_32BIT(0x605C)
+#define ACT_CTRL_0				AP1302_REG_16BIT(0x6066)
+#define ACT_CTRL_1				AP1302_REG_16BIT(0x6068)
 
 struct ap1302_device;
 
@@ -2553,6 +2556,14 @@ static int ap1302_load_firmware(struct ap1302_device *ap1302)
 	 * PIPE FREQ = 112 MHZ (564 from PLL0 /5) */
 	ap1302_write(ap1302, AP1302_PREVIEW_DIV_IPIPE, 0x00010008, NULL);
 
+	if (!strcmp(ap1302->sensor_info->model, AR1335_MODEL)) {
+		/* Set Actuator absolute milisecond time to 5ms */
+		ap1302_write(ap1302, ACT_ABS_MTIME, 0x00001388, NULL);
+		/* Initialise Actuator controls */
+		ap1302_write(ap1302, ACT_CTRL_0, 0x00004C00, NULL);
+		ap1302_write(ap1302, ACT_CTRL_1, 0x00000110, NULL);
+	}
+
 	/* Check again if PLL is locked aftaer adjusting Host interface MIPI data rate of AP1302 */
 	for (cnt = 0 ; !ret && (cnt < 10) ; cnt++) {
 		msleep(2);
-- 
2.17.1

